### PR TITLE
Add extend ability to query external service for autoscaler

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve? 
Close https://github.com/pingcap/tidb-operator/pull/2235

Add extend ability to query external service for autoscaler to let community user could use their own strategy to adjust the auto-scaling.

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
Make `TidbClusterAutoScaler` have ther external strategy ability
```

### CODE REVIEW

**Comment by Yisaer** [comment]
@cofyc @DanielZhangQD @weekface  PTAL

File: pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go [hunk 1]
```diff
@@ -198,3 +203,26 @@ type MetricsStatus struct {
...
 	ThresholdValue string `json:"thresholdValue"`
 }
+
+// +k8s:openapi-gen=true
+// ExternalEndpoint describe the external service enpoint
+// who provide the tikv/tidb auto-scaling recommended replicas ability
+type ExternalEndpoint struct {
+	// Host indicates the external service's host
+	Host string `json:"host"`
+	// Port indicates the external service's post
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go:213]:
```suggestion
	// Port indicates the external service's port
```

File: pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go [hunk 2]
```diff
@@ -198,3 +203,26 @@ type MetricsStatus struct {
...
+	// to communicate to the external service
+	// +optional
+	TlsSecretRef *SecretRef `json:"tlsSecret,omitempty"`
+}
+
+// +k8s:openapi-gen=true
+// SecretRef indicates to secret ref
+type SecretRef struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go:227]: If not set, we can take the namespace of tidb-controller-manager.

File: pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go [hunk 3]
```diff
@@ -135,6 +135,11 @@ type BasicAutoScalerSpec struct {
 	// If not set, the default value is 5.
 	// +optional
 	ScaleInThreshold *int32 `json:"scaleInThreshold,omitempty"`
+
+	// ExternalEndpoint make auto-scaler controller could query the external service by http/https to
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go:139]:
```suggestion
	// ExternalEndpoint makes the auto-scaler controller able to query the external service
```

File: pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go [hunk 4]
```diff
@@ -135,6 +135,11 @@ type BasicAutoScalerSpec struct {
...
+	// fetch the result of the recommended replicas for tikv/tidb
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go:140]:
```suggestion
	// to fetch the recommended replicas for TiKV/TiDB
```

File: pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go [hunk 5]
```diff
@@ -198,3 +203,26 @@ type MetricsStatus struct {
 	// TargetValue indicates the threshold value for this metrics in auto-scaling
 	ThresholdValue string `json:"thresholdValue"`
 }
+
+// +k8s:openapi-gen=true
+// ExternalEndpoint describe the external service enpoint
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go:208]:
```suggestion
// ExternalEndpoint describes the external service endpoint
```

File: pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go [hunk 6]
```diff
@@ -198,3 +203,26 @@ type MetricsStatus struct {
...
+// who provide the tikv/tidb auto-scaling recommended replicas ability
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go:209]:
```suggestion
// which provides the ability to get the tikv/tidb auto-scaling recommended replicas
```

File: pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go [hunk 7]
```diff
@@ -198,3 +203,26 @@ type MetricsStatus struct {
...
+type ExternalEndpoint struct {
+	// Host indicates the external service's host
+	Host string `json:"host"`
+	// Port indicates the external service's post
+	Port int32 `json:"port"`
+	// Path indicates the external service's path
+	Path string `json:"path"`
+	// TlsSecretRef indicates the Secret which stored the tls configuration. If enabled, operator would use https
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go:217]:
```suggestion
	// TLSSecret indicates the Secret which stores the TLS configuration. If set, the operator will use https
```

File: pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go [hunk 8]
```diff
@@ -198,3 +203,26 @@ type MetricsStatus struct {
...
+	// to communicate to the external service
+	// +optional
+	TlsSecretRef *SecretRef `json:"tlsSecret,omitempty"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go:220]:
```suggestion
	TLSSecret *SecretRef `json:"tlsSecret,omitempty"`
```

File: pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go (hunk 7 above)
Comment (DanielZhangQD) [reply pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go:217]: The keys in the secret should be the fixed format, e.g. tls.crt, right? If yes, we need to make it clear here.

File: pkg/autoscaler/autoscaler/query/external.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+	if err != nil {
+		return -1, err
+	}
+	resp := &ExternalResponse{}
+	err = json.Unmarshal(bytes, resp)
+	if err != nil {
+		return -1, err
+	}
+	if resp.Name != tc.Name || resp.Namespace != tc.Namespace || resp.Type != memberType.String() {
+		return -1, fmt.Errorf("external endpoint return wrong info,%v", resp)
```
Comment (DanielZhangQD) [inline pkg/autoscaler/autoscaler/query/external.go:48]:
```suggestion
		return -1, fmt.Errorf("external endpoint returns unexpected info, get %#v, expect %s/%s/%s", resp, tc.Namespace, tc.Name, memberType)
```

File: pkg/autoscaler/autoscaler/query/external.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+	}
+	return resp.RecommendedReplicas, nil
+}
+
+func sendRequest(tc *v1alpha1.TidbCluster, memberType v1alpha1.MemberType, endpoint *v1alpha1.ExternalEndpoint, kubecli kubernetes.Interface) ([]byte, error) {
+	schema := "http"
```
Comment (DanielZhangQD) [inline pkg/autoscaler/autoscaler/query/external.go:54]:
```suggestion
	scheme := "http"
```

File: pkg/autoscaler/autoscaler/query/external.go [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+		tlsConfig, err := loadTlsConfig(endpoint, kubecli)
+		if err != nil {
+			return nil, err
+		}
+		tr := &http.Transport{
+			TLSClientConfig: tlsConfig,
+		}
+		client = &http.Client{Transport: tr}
+	} else {
+		client = &http.Client{}
```
Comment (DanielZhangQD) [inline pkg/autoscaler/autoscaler/query/external.go:67]: Set a timeout for the client?

File: pkg/autoscaler/autoscaler/query/external.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+	r, err := client.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	bytes, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return nil, err
+	}
+	if r.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("query from external endpoint[%s] failed, response:%v", url, string(bytes))
```
Comment (DanielZhangQD) [inline pkg/autoscaler/autoscaler/query/external.go:79]:
```suggestion
		return nil, fmt.Errorf("query from external endpoint [%s] failed, response: %v, status code: %v", url, string(bytes), r.StatusCode)
```

File: pkg/autoscaler/autoscaler/query/external.go [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+	}
+	return bytes, nil
+}
+
+func loadTlsConfig(endpoint *v1alpha1.ExternalEndpoint, kubecli kubernetes.Interface) (*tls.Config, error) {
```
Comment (DanielZhangQD) [inline pkg/autoscaler/autoscaler/query/external.go:84]:
```suggestion
func loadTLSConfig(endpoint *v1alpha1.ExternalEndpoint, kubecli kubernetes.Interface) (*tls.Config, error) {
```

File: pkg/autoscaler/autoscaler/tidb_autoscaler.go [hunk 14]
```diff
@@ -39,17 +41,25 @@ func (am *autoScalerManager) syncTiDB(tc *v1alpha1.TidbCluster, tac *v1alpha1.Ti
...
+	if tac.Spec.TiDB.ExternalEndpoint == nil {
+		instances := filterTidbInstances(tc)
+		targetReplicas, err = calculateTidbMetrics(tac, sts, instances)
+		if err != nil {
+			return err
+		}
+	} else {
+		targetReplicas, err = query.ExternalService(tc, v1alpha1.TiDBMemberType, tac.Spec.TiDB.ExternalEndpoint, am.kubecli)
+		if err != nil {
+			klog.Errorf("tac[%s/%s] 's externalEndpoint meet error, err:%v", tac.Namespace, tac.Name, err)
```
Comment (DanielZhangQD) [inline pkg/autoscaler/autoscaler/tidb_autoscaler.go:54]:
```suggestion
			klog.Errorf("tac [%s/%s] 's query to the external endpoint got error: %v", tac.Namespace, tac.Name, err)
```

File: pkg/autoscaler/autoscaler/tikv_autoscaler.go [hunk 15]
```diff
@@ -41,15 +43,24 @@ func (am *autoScalerManager) syncTiKV(tc *v1alpha1.TidbCluster, tac *v1alpha1.Ti
...
+	var targetReplicas int32
+	if tac.Spec.TiDB.ExternalEndpoint == nil {
+		targetReplicas, err = calculateTikvMetrics(tac, sts, instances)
+		if err != nil {
+			return err
+		}
+	} else {
+		targetReplicas, err = query.ExternalService(tc, v1alpha1.TiKVMemberType, tac.Spec.TiKV.ExternalEndpoint, am.kubecli)
+		if err != nil {
+			klog.Errorf("tac[%s/%s] 's externalEndpoint meet error, err:%v", tac.Namespace, tac.Name, err)
```
Comment (DanielZhangQD) [inline pkg/autoscaler/autoscaler/tikv_autoscaler.go:55]: ditto

File: pkg/autoscaler/autoscaler/query/external.go (hunk 11 above)
Comment (Yisaer) [reply pkg/autoscaler/autoscaler/query/external.go:67]: updated.

File: pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go (hunk 2 above)
Comment (Yisaer) [reply pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go:227]: I think it is not necessary and didn't make sense.

**Comment by sunzhuohang** [comment]
LGTM

File: pkg/autoscaler/autoscaler/query/external.go [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+	}
+	scheme := "http"
+	if endpoint.TLSSecret != nil {
+		scheme = "https"
+	}
+	url := fmt.Sprintf("%s://%s:%d%s?name=%s&namespace=%s&type=%s", scheme, endpoint.Host, endpoint.Port, endpoint.Path, tc.Name, tc.Namespace, memberType.String())
+	r, err := client.Get(url)
+	if err != nil {
+		return nil, err
+	}
```
Comment (weekface) [inline pkg/autoscaler/autoscaler/query/external.go:71]:
```suggestion
	}
	defer r.Body.Close()
```

**Pull request merged**